### PR TITLE
Jupiter 10.11.0 plat 2857 1

### PR DESCRIPTION
**Pull request by noam-arad**
Merge so we can start integrating parts.
Still needs calculate score.

### CODE REVIEW

File: alpha/lib/enums/QuizUserEntryStatus.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+/**
+ * @package Core
+ * @subpackage model.enum
+ */
+interface QuizUserEntryStatus extends UserEntryStatus {
```
Comment (tan-tan-kanarek) [inline alpha/lib/enums/QuizUserEntryStatus.php]: Please move this class into the Quiz plugin and make it IKalturaPluginEnum.

File: alpha/lib/enums/UserEntryType.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+/**
+ * @package Core
+ * @subpackage model.enum
+ */
+interface userEntryType extends BaseEnum
+{
+	const KALTURA_QUIZ_USER_ENTRY =1;
```
Comment (tan-tan-kanarek) [inline alpha/lib/enums/UserEntryType.php]: This enum should be empty and the Quiz value should be added by the Quiz plugin.

File: alpha/lib/model/QuizUserEntry.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+/**
+ * @package Core
+ * @subpackage model
+ */
+class QuizUserEntry extends UserEntry{
```
Comment (tan-tan-kanarek) [inline alpha/lib/model/QuizUserEntry.php]: Please move it to the Quiz plugin.

File: alpha/lib/model/UserEntryPeer.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * You should add additional methods to this class to meet the
+ * application requirements.  This class will only be generated as
+ * long as it does not already exist in the output directory.
+ *
+ * @package Core
+ * @subpackage model
+ */
+class UserEntryPeer extends BaseUserEntryPeer {
+
+	const QUIZ_OM_CLASS = 'QuizUserEntry';
```
Comment (tan-tan-kanarek) [inline alpha/lib/model/UserEntryPeer.php]: Please remove this constant, the word "quiz" shouldn't appear anywhere outside the Quiz plugin.

File: plugins/cue_points/quiz/lib/services/QuizUserEntryService.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
+<?php
```
Comment (tan-tan-kanarek) [inline plugins/cue_points/quiz/lib/services/QuizUserEntryService.php]: services directory should be directly under the plugin directory, not under lib, see other plugins.

File: plugins/cue_points/quiz/lib/services/QuizUserEntryService.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+		$userEntry = new KalturaQuizUserEntry();
+		$userEntry->fromObject($dbUserEntry, $this->getResponseProfile());
+		return $userEntry;
+	}
+
+	/**
+	 * @return int
+	 */
+	protected function calculateScore()
```
Comment (tan-tan-kanarek) [inline plugins/cue_points/quiz/lib/services/QuizUserEntryService.php]: I think this method should be on the QuizUserEntry object.

File: deployment/permissions/service.quiz.quizuserentry.ini [hunk 7]
```diff
@@ -0,0 +1,5 @@
+[action_permission_items]
```
Comment (tan-tan-kanarek) [inline deployment/permissions/service.quiz.quizuserentry.ini]: I don't think this file is needed, the permissions already granted in service.userentry.ini.

File: deployment/permissions/service.userentry.ini [hunk 8]
```diff
@@ -0,0 +1,51 @@
+[action_permission_items]
+permissionItem1.service = userentry
+permissionItem1.action = add
+permissionItem1.partnerId = 0
+permissionItem1.param3 =
+permissionItem1.param4 =
+permissionItem1.param5 =
+permissionItem1.tags =
+permissionItem1.permissions = BASE_USER_SESSION_PERMISSION
```
Comment (tan-tan-kanarek) [inline deployment/permissions/service.userentry.ini]: I think you should add here also WIDGET_SESSION_PERMISSION.

File: deployment/permissions/service.userentry.ini [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+permissionItem5.param3 =
+permissionItem5.param4 =
+permissionItem5.param5 =
+permissionItem5.tags =
+permissionItem5.permissions = BASE_USER_SESSION_PERMISSION
+
+permissionItem6.service = userentry
+permissionItem6.action = submitQuiz
+permissionItem6.partnerId = 0
+permissionItem6.permissions = BASE_USER_SESSION_PERMISSION
```
Comment (tan-tan-kanarek) [inline deployment/permissions/service.userentry.ini]: I think you should add here also WIDGET_SESSION_PERMISSION.

File: api_v3/services/UserEntryService.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	 */
+	public function addAction(KalturaUserEntry $userEntry)
+	{
+		$dbUserEntry = $userEntry->toInsertableObject(null, array('type'));
+		$userId = $userEntry->userId;
+		if ($userId == 0)
+		{
+			$userId = kCurrentContext::$ks_kuser;
+		}
+		$dbUserEntry->setKuserId($userId);
```
Comment (tan-tan-kanarek) [inline api_v3/services/UserEntryService.php]: Please implement all these setters inside the KalturaUserEntry object.
user id and status in the toInsertableObject and type in the constructor.

File: api_v3/services/UserEntryService.php [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	 * @throws KalturaAPIException
+	 */
+	public function updateAction($id, KalturaUserEntry $userEntry)
+	{
+		$dbUserEntry = UserEntryPeer::retrieveByPK($id);
+		if (!$dbUserEntry)
+			throw new KalturaAPIException(KalturaErrors::INVALID_OBJECT_ID, $id);
+
+		$dbUserEntry = $userEntry->toUpdatableObject($dbUserEntry);
+		$dbUserEntry->setUpdatedat(time());
```
Comment (tan-tan-kanarek) [inline api_v3/services/UserEntryService.php]: setUpdatedat is called automatically when you call save().

File: api_v3/services/UserEntryService.php [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+			$response->totalCount = 0;
+			return $response;
+		}
+		if (!$filter)
+		{
+			$filter = new KalturaUserEntryFilter();
+		}
+		$c = new Criteria();
+
+		$userEntryFilter = new UserEntryFilter();
```
Comment (tan-tan-kanarek) [inline api_v3/services/UserEntryService.php]: Please move the logic into KalturaUserEntryFilter, see KalturaAccessControlProfileFilter::KalturaUserEntryFilter as a simple example.

File: api_v3/lib/types/KalturaUserEntryFilter.php [hunk 13]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+/**
+ * @package api
+ * @subpackage filters
+ */
+class KalturaUserEntryFilter extends KalturaRelatedFilter
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/KalturaUserEntryFilter.php]: This class should be auto-generated and should extend KalturaUserEntryBaseFilter.

File: api_v3/lib/types/KalturaUserEntry.php [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+	 * @return KalturaUserEntry
+	 *
+	 * MOVE FUNCTION to quiz plugin
+	 */
+	public static function getInstanceByType ($type)
+	{
+		$obj = null;
+		switch ($type) {
+			case KalturaUserEntryType::KALTURA_QUIZ_USER_ENTRY:
+				$obj = new KalturaQuizUserEntry();
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/KalturaUserEntry.php]: Remove the Quiz from here, the KalturaQuizUserEntry should be loaded by the plugin.

File: api_v3/lib/types/KalturaUserEntry.php [hunk 15]
```diff
@@ -0,0 +1,106 @@
+<?php
+/**
+ * @package api
+ * @subpackage objects
+ * @abstract
+ */
+abstract class KalturaUserEntry extends KalturaObject
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/KalturaUserEntry.php]: entryId and userId should be @insertonly.
partnerId should be @readonly.

File: deployment/permissions/service.quiz.quizuserentry.ini (hunk 7 above)
Comment (noam-arad) [inline deployment/permissions/service.quiz.quizuserentry.ini]: From my test that is not so, I also looked at a similar example of aliasAction

File: plugins/cue_points/quiz/lib/services/QuizUserEntryService.php (hunk 5 above)
Comment (noam-arad) [inline plugins/cue_points/quiz/lib/services/QuizUserEntryService.php]: I also thought that, maybe @MichalRadwantzor knows why it is the way it is

Comment (MichalRadwantzor) [inline plugins/cue_points/quiz/lib/services/QuizUserEntryService.php]: @noam-arad  my bad, I fixed now.

**Pull request merged**